definitions/npm/firebase_v2.4.2/flow_v0.34.x-/firebase_v2.4.2.js

```diff
@@ -0,0 +1,247 @@
+    user: ?User,
+    credential: ?AuthCredential,
+    operationType: ?string,
+  };
+
+  declare class Database {
+    app: App,
+    goOffline(): void,
+    goOnline(): void,
+    ref(path: ?string): Reference,
+    refFromURL(url: string): Reference,
+  }
+
+  declare class DataSnapshot {
+    key: ?string,
+    ref: Reference,
+    child(path: string): DataSnapshot,
+    exists(): boolean,
+    exportVal(): mixed,
+    forEach(action: (DataSnapshot) => void): boolean,
+    getPriority(): string|number|null,
+    hasChild(path: string): boolean,
+    hasChildren(): boolean,
+    numChildren(): number,
+    toJSON(): {},
+    val(): mixed,
+  }
+
+  declare class OnDisconnect {
+    cancel(onComplete: ?OnCompleteCallback): Promise<void>,
+    remove(onComplete: ?OnCompleteCallback): Promise<void>,
+    set(value: mixed, onComplete: ?OnCompleteCallback): Promise<void>,
+    setWithPriority(value: string, priority: number|string|null, onComplete: ?OnCompleteCallback): Promise<void>,
+    update(values: {}, onComplete: ?OnCompleteCallback): Promise<void>,
+  }
+
+  declare class Query {
+    ref: Reference,
+    endAt(value: number|string|boolean|null, key: ?string): Query,
+    equalTo(value: number|string|boolean|null, key: ?string): Query,
+    isEqual(other: Query): boolean,
+    limitToFirst(limit: number): Query,
+    limitToLast(limit: number): Query,
+    off(eventType: string, callback: ?(DataSnapshot, ?string) => void, context: ?{}): void,
+    on(eventType: string, callback: ?(DataSnapshot, ?string) => void, cancelCallbackOrConetxt?: {}|(Error) => void, context: ?{}): (DataSnapshot, ?string) => void,
+    once(eventType: string, callback: ?(DataSnapshot, ?string) => void, cancelCallbackOrContext?: {}|(Error) => void, context: ?{}): Promise<DataSnapshot>,
+    orderByChild(path: string): Query,
+    orderByKey(): Query,
+    orderByPriority(): Query,
+    orderByValue(): Query,
+    startAt(value: number|string|boolean|null, key: ?string): Query,
+    toJSON(): {},
+    toString(): string,
+  }
+
+  declare type OnCompleteCallback = (?Error) => void;
+
+  declare class Reference extends Query {
+    key: ?string,
+    parent: ?Reference,
+    root: Reference,
+
+    child(path: string): Reference,
+    onDisconnect(): OnDisconnect,
+    push(value: ?mixed, onComplete: ?OnCompleteCallback): ThenableReference<void>,
+    remove(onComplete: ?OnCompleteCallback): ThenableReference<void>,
+    set(value: mixed, onComplete: ?OnCompleteCallback): Promise<void>,
+    setPriority(priority: string|number|null, onComplete: ?OnCompleteCallback): Promise<void>,
+    setWithPriority(newVal: mixed, newPriority: string|number|null, onComplete: ?OnCompleteCallback): Promise<void>,
+    transaction(transactionUpdate: (mixed) => void, onComplete: ?(?Error, boolean, ?DataSnapshot) => void, applyLocally: ?boolean): Promise<{committed: boolean, snapshot: ?DataSnapshot,}>,
+    update(values: {}, onComplete: ?OnCompleteCallback): Promise<void>,
```

Comment:
Instead of `value: mixed` (in `push` and `set`), and `values: {}` (in `update`), I've been using:

```
    declare type Value = string | number | boolean | Array<any> | Object;
    declare type Object = {| [key: string | number]: Value |}
…
    push(Object, ?OnCompleteCallback): Reference;
    set(Value | null, ?OnCompleteCallback): Promise<void>;
    setPriority(Priority, ?OnCompleteCallback): Promise<void>;
    update(Object, ?OnCompleteCallback): Promise<void>;
```

This should protect against a non-storable value such as a function being passed to one of these methods.

Disclaimer: I've only been using flow-type with Firebase on a small project for a short time, and haven't tested whether my version runs into problems.